jupyter_kernel_test/__init__.py

```diff
@@ -188,11 +204,111 @@ def test_display_data(self):
             raise SkipTest
 
         for sample in self.code_display_data:
-            self.flush_channels()
-            reply, output_msgs = self.execute_helper(sample['code'])
+            with self.subTest(code=sample['code']):
+                self.flush_channels()
+                reply, output_msgs = self.execute_helper(sample['code'])
+
+                self.assertEqual(reply['content']['status'], 'ok')
+
+                self.assertGreaterEqual(len(output_msgs), 1)
+                self.assertEqual(output_msgs[0]['msg_type'], 'display_data')
+                self.assertIn(sample['mime'], output_msgs[0]['content']['data'])
+
+    def test_silent(self):
+        self.flush_channels()
+
+        reply, output_msgs = self.execute_helper("", silent=True)
+        self.assertEqual(reply['content']['status'], 'ok')
+        self.assertEqual(len(output_msgs), 0)
+        # test that execution_count has not been incremented?
+
+    # this should match one of the values in code_execute_result
+    code_history_pattern = ""
+
+    def history_helper(self, execute_first, timeout=TIMEOUT, **histargs):
+        self.flush_channels()
 
-            self.assertEqual(reply['content']['status'], 'ok')
+        for code in execute_first:
+            reply, output_msgs = self.execute_helper(code)
+
+        self.flush_channels()
+        msg_id = self.kc.history(**histargs)
+
+        reply = self.kc.get_shell_msg(timeout=timeout)
+        validate_message(reply, 'history_reply', msg_id)
+
+        return reply
+
+    def test_history(self):
+        if not self.code_execute_result:
+            raise SkipTest
```

Comment:
I originally had separate tests for the different history modes, and bundled them into a single function with subtests to reduce code duplication. Possibly separating them and adding multiple flags would be better here? I assume that given that history is (for now) documented in the messaging protocol you're happy to carry tests for it.